### PR TITLE
Populate a per-repository map of normalized version tags

### PR DESCRIPTION
**Pull request by andrewpollock**
This will facilitate exact and fuzzy matches on CPE versions in CVE records.

### CODE REVIEW

File: vulnfeeds/git/repository.go [hunk 1]
```diff
@@ -26,6 +27,11 @@ func (v Versions) Swap(i, j int)      { v[i], v[j] = v[j], v[i] }
 // RepoTagsMap acts as a cache for RepoTags results, keyed on the repo's URL.
 type RepoTagsMap map[string]Versions
 
+// NormalizedRepoTagsMap allows for looking up a repo's tags by normalized values.
+// A composite struct key of repo and normalized tag is not used to allow for
+// fuzzy matches on all available normalized versions for a given repo
+type NormalizedRepoTagsMap map[string]map[string]Version
```
Comment (oliverchang) [inline vulnfeeds/git/repository.go:33]: nit: this is a little hard to understand. We should either document the expected keys/nesting structure. i.e.

```
repo URL -> normalized verison -> original unnormalized version
```

Or restructure some of these data structures to have some better abstraction. Something like: 

```go
// NormalizedVersions provides mappings from normalized versions to their original tag value.
type NormalizedVersions struct {
  OriginalVersion map[string]string
}

// RepoNormalizedVersionsCache is a cache allowing for lookups of normalized versions per repo. 
type RepoNormalizedVersionsCache struct {
  Repo map[string]NormalizedVersions
}

// Usage
normalizedVersionsCache.Repo[url].OriginalVersion[v]
```

File: vulnfeeds/git/repository.go [hunk 2]
```diff
@@ -63,6 +69,29 @@ func RepoTags(repoURL string, repoTagsCache *RepoTagsMap) (versions Versions, e
 	return versions, nil
 }
 
+// NormalizeRepoTags populates a durable mapping of tags to Versions for lookup by normalized tag.
+func NormalizeRepoTags(repoURL string, normalizedRepoTags NormalizedRepoTagsMap, repoTagsCache *RepoTagsMap) (e error) {
```
Comment (oliverchang) [inline vulnfeeds/git/repository.go:73]: it's not clear what's the output of this function, as it's the second argument here. Can we return the NormalizedRepoTagsMap map instead?

File: vulnfeeds/git/repository.go [hunk 3]
```diff
@@ -8,6 +8,7 @@ import (
 	"github.com/go-git/go-git/v5/config"
 	"github.com/go-git/go-git/v5/plumbing/transport"
 	"github.com/go-git/go-git/v5/storage/memory"
+	"github.com/google/osv/vulnfeeds/cves"
```
[the remark this answers is not in the document]
Comment (andrewpollock) [reply vulnfeeds/git/repository.go:12]: Done in[a637ae8](https://github.com/google/osv.dev/pull/1057/commits/a637ae8da185a651320f26ac49cd72051220d183)

File: vulnfeeds/git/repository.go (hunk 2 above)
Comment (oliverchang) [reply vulnfeeds/git/repository.go:74]: I mean the bits that did get computed (not NormalizedRepoTagsMap as a whole): 

i.e. either the `map[string]Version` as it is now or `NormalizedVersions` as I suggested above with the refactoring.

Comment (andrewpollock) [reply vulnfeeds/git/repository.go]: Yep, done in [6907948](https://github.com/google/osv.dev/pull/1057/commits/69079488e660ad00d4d6971eb2bb36ca0a7dc42c)

File: vulnfeeds/git/repository.go (hunk 1 above)
Comment (andrewpollock) [reply vulnfeeds/git/repository.go]: Thanks, I've given this a thorough overhaul (which conveniently surfaced some bugs in both my understanding of maps in Go and my caching logic). PTAL.

File: vulnfeeds/git/repository.go [hunk 4]
```diff
@@ -8,31 +8,46 @@ import (
...
+type NormalizedTag struct {
+	OriginalTag string
+	Commit      string
+}
 
 // RepoTagsMap acts as a cache for RepoTags results, keyed on the repo's URL.
-type RepoTagsMap map[string]Versions
+// repo URL -> Tag -> GitTag
+type RepoTagsMap struct {
+	Tag           map[string]Tag
```
Comment (oliverchang) [inline vulnfeeds/git/repository.go:37]: nit: can you add some short comments on the expected key for these? It's not obvious from the code itself.

File: vulnfeeds/git/repository.go [hunk 5]
```diff
@@ -8,31 +8,46 @@ import (
...
+func (t Tags) Swap(i, j int)      { t[i], t[j] = t[j], t[i] }
+
+type NormalizedTag struct {
+	OriginalTag string
+	Commit      string
+}
 
 // RepoTagsMap acts as a cache for RepoTags results, keyed on the repo's URL.
-type RepoTagsMap map[string]Versions
+// repo URL -> Tag -> GitTag
```
Comment (oliverchang) [inline vulnfeeds/git/repository.go:35]: Is this comment still relevant?

**Review by oliverchang — APPROVED** [review]
LGTM with some comment nits.

File: vulnfeeds/git/repository.go (hunk 4 above)
Comment (andrewpollock) [reply vulnfeeds/git/repository.go]: Improved in [95c7d05](https://github.com/google/osv.dev/pull/1057/commits/95c7d0521ed2ea24906a34903e6f2469d4fead2d)

File: vulnfeeds/git/repository.go (hunk 5 above)
Comment (andrewpollock) [reply vulnfeeds/git/repository.go]: No, thanks for highlighting. Fixed in [95c7d05](https://github.com/google/osv.dev/pull/1057/commits/95c7d0521ed2ea24906a34903e6f2469d4fead2d)

**Pull request merged**